lib/compress/zstd_lazy.c

```diff
@@ -615,52 +908,128 @@ size_t ZSTD_HcFindBestMatch_generic (
+
+        /* Generate a "matches" bitfield. The nth bit == 1 if the newly-computed "tag" matches the hash at the nth
+           position in a row of the tagTable */
+#ifdef __SSE2__
+        if (rowEntries == 16) {
+            ZSTD_Vec128 hashes = ZSTD_Vec128_read(tagRow + kHashOffset);
+            ZSTD_Vec128 hash1  = ZSTD_Vec128_set8(tag);
+            ZSTD_Vec128 cmpeq  = ZSTD_Vec128_cmp8(hashes, hash1);
+            matches            = ZSTD_Vec128_mask8(cmpeq);
+        } else if (rowEntries == 32) {
+            ZSTD_Vec256 hashes = ZSTD_Vec256_read(tagRow + kHashOffset);
+            ZSTD_Vec256 hash1  = ZSTD_Vec256_set8(tag);
+            ZSTD_Vec256 cmpeq  = ZSTD_Vec256_cmp8(hashes, hash1);
+            matches            = ZSTD_Vec256_mask8(cmpeq);
         } else {
-            const BYTE* const match = dictBase + matchIndex;
-            assert(match+4 <= dictEnd);
-            if (MEM_read32(match) == MEM_read32(ip))   /* assumption : matchIndex <= dictLimit-4 (by table construction) */
-                currentMl = ZSTD_count_2segments(ip+4, match+4, iLimit, dictEnd, prefixStart) + 4;
+            assert(0);
         }
+#else
+        if (rowEntries == 16) {
+            ZSTD_Vec128 hashes = {0};
+            ZSTD_Vec128 vec = {0};
+            ZSTD_Vec128_scalarRead(tagRow + kHashOffset, hashes);
+            ZSTD_Vec128_scalarSet(tag, vec);
+            matches  = ZSTD_Vec128_scalarCmp(hashes, vec);
+        } else if (rowEntries == 32) {
+            ZSTD_Vec256 hashes = {0};
+            ZSTD_Vec256 vec = {0};
+            ZSTD_Vec256_scalarRead(tagRow + kHashOffset, hashes);
+            ZSTD_Vec256_scalarSet(tag, vec);
+            matches  = ZSTD_Vec256_scalarCmp(hashes, vec);
+        }
+#endif /* __SSE2__ */
+
+        /* Each row is a circular buffer beginning at the value of "head". So we must rotate the "matches" bitfield
+           to match up with the actual layout of the entries within the hashTable */
+        matches = ZSTD_VecMask_rotateRight(matches, head, rowEntries);
 
-        /* save best solution */
-        if (currentMl > ml) {
-            ml = currentMl;
-            *offsetPtr = curr - matchIndex + ZSTD_REP_MOVE;
-            if (ip+currentMl == iLimit) break; /* best possible, avoids read overflow on next attempt */
+        /* Cycle through the matches and prefetch */
+        for (; (matches > 0) && (nbAttempts > 0); --nbAttempts, matches &= (matches - 1)) {
+            U32 const matchPos = (head + ZSTD_VecMask_next(matches)) & rowMask;
+            U32 const matchIndex = row[matchPos];
+            if (matchIndex < lowLimit)
+                break;
+            if (shouldPrefetch) {
+                if ((dictMode != ZSTD_extDict) || matchIndex >= dictLimit) {
+                    PREFETCH_L1(base + matchIndex);
+                } else {
+                    PREFETCH_L1(dictBase + matchIndex);
+                }
+            }
+            matchBuffer[numMatches++] = matchIndex;
```

Comment:
Have you tried inserting the scalar fallback here instead?

In the SSE version we iterate over the `matches` mask.

In the scalar version, we could iterate over the tags.
```c++
for (U32 pos = 0; pos < 16; ++pos) {
  U32 const matchPos = (head + pos) & rowMask;
  U32 const matchIndex = row[matchPos];
  if (tagRow[matchPos] != tag) continue;
  if (matchIndex < lowLimit) break;
  if (nbAttempts == 0) break;
  --nbAttempts;

  // Rest is the same
```

You could also try a cmov-style version to try to avoid branches. Though, who knows if the architecture supports cmov or not, since this code won't be executing on x86.

```c++
for (U32 pos = 0; pos < 16; ++pos) {
  U32 const matchPos = (head + pos) & rowMask;
  U32 const matchIndex = row[matchPos];
  int const tagMatches = (tagRow[matchPos] == tag) ? 1 : 0;
  if (matchIndex < lowLimit) break;
  if (nbAttempts == 0) break;
  nbAttempts -= tagMatches;

  // Rest is the same except don't ++ numMatches & don't prefetch b/c it is probably disabled anyways because not x86

  numMatches += tagMatches;
```

Its probably not right to be measuring performance on x86-64, since we know for a fact that the scalar code won't be executing on x86-64.

You could try executing in 32-bit mode (`-m32`). That would give a slightly more interesting comparison. And/or you could measure on ARM, though that is harder to set up. Either of those would be a more interesting comparison than x86-64 speed.

Reply:
Definitely an interesting idea, and a more natural way to go about it - I'll give it a shot!

Reply:
After some more measurements, it actually doesn't seem like either of these offers a speed improvement compared to 32-bit no SIMD. On very small files it seems to provides some benefit, but we don't use the row-based matchfinder for those small files at that point anyways.